File: Data Structures & Algorithms/subsets/submission-4.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def createSubsets(i, curr):
            if i == len(nums):
                res.append(curr.copy())
                return

            curr.append(nums[i])
            createSubsets(i + 1, curr)

            curr.pop()
            createSubsets(i + 1, curr)

        createSubsets(0, [])
        return res

File: Data Structures & Algorithms/combinations/submission-4.ts
class Solution {
    /**
     * @param {number} n
     * @param {number} k
     * @return {number[][]}
     */
    combine(n: number, k: number): number[][] {
        const res: number[][]  = []

        const backtracking = (i: number, curr: number[]): void => {
            if (i > n) {
                if (curr.length === k) {
                    res.push([...curr])
                }
                return
            }

            curr.push(i)
            backtracking(i + 1, curr)
            curr.pop()
            backtracking(i + 1, curr)
        }

        backtracking(1, [])
        return res
    }
}
